torch/csrc/jit/script/compiler.cpp

```diff
@@ -34,18 +34,122 @@ using ListAttributeMap = std::unordered_map<
     std::string,
     std::pair<const std::vector<double>, std::string>>;
 
+// Keep track of environment as we descend down nested control
+// structures.
+struct Environment {
+  Environment(Block* b, std::shared_ptr<Environment> next = nullptr)
+      : b(b), next(next) {}
+
+  std::vector<std::string> captured_inputs;
+  ValueTable value_table;
+  Block* b;
+
+  std::shared_ptr<Environment> next;
+
+  Value* findInThisFrame(const std::string& name) {
+    if (value_table.count(name)) {
+      return value_table.at(name);
+    }
+    return nullptr;
+  }
+
+  Value* findInParentFrame(const std::string& name) {
+    for (auto runner = next; runner; runner = runner->next) {
+      if (runner->value_table.count(name)) {
+        return runner->value_table.at(name);
+      }
+    }
+    return nullptr;
+  }
+
+  Value* createCapturedInput(const std::string& name) {
+    // Create the input
+    Value* new_input = b->addInput();
+
+    // Associate this name with this value
+    value_table[name] = new_input;
+
+    // List as a positional input
+    captured_inputs.push_back(name);
+
+    return new_input;
+  }
+
+  Symbol getBlockOwningKind() {
+    Symbol owning_kind = Symbol();
+    if (b->owningNode()) {
+      owning_kind = b->owningNode()->kind();
+    }
+    return owning_kind;
+  }
+
+  void setVar(const std::string& name, Value* value) {
+    if (!findInThisFrame(name) && findInParentFrame(name) &&
+        getBlockOwningKind() == Symbol("Loop"))
+      createCapturedInput(name);
+    value_table[name] = value;
+  }
+
+  Value* getVar(const Ident& ident) {
+    return getVar(ident.name(), ident);
+  }
+
+  Value* getVar(const std::string& ident, const TreeView& tv) {
+    Value* retval = findInThisFrame(ident);
+
+    if (!retval && (retval = findInParentFrame(ident)) &&
+        getBlockOwningKind() == Symbol("Loop"))
+      retval = createCapturedInput(ident);
+
+    if (!retval) {
+      throw ErrorReport(tv) << "undefined value " << ident;
+    }
+
+    return retval;
+  }
+
+  // Given that after emitting statements in a block, we've added block inputs
+  // for all value references and assignments, delete inputs for which there was
+  // no assignment, only references.
+  void deleteExtraInputs() {
+    std::vector<size_t> inputs_to_delete;
+    int i = 0;
+    for (const auto& x : captured_inputs) {
+      if (b->inputs()[i] == value_table[x]) {
+        inputs_to_delete.push_back(i);
+      }
+      i++;
+    }
+
+    for (auto ritr = inputs_to_delete.rbegin(); ritr != inputs_to_delete.rend();
+         ++ritr) {
+      auto name = captured_inputs[*ritr];
+      Value* v = value_table[name];
+      Value* orig = findInParentFrame(name);
+      // Replace all matching node inputs with original value
+      // from an enclosing scope
+      v->replaceAllUsesWith(orig);
+
+      // Actually remove the input
+      b->eraseInput(*ritr);
+      captured_inputs.erase(captured_inputs.begin() + *ritr);
+    }
+  }
+};
```

Comment:
If we want to maintain compatibility with Python semantics, we can’t make control flow create new frames. In Python only a function call creates a new one, and all variables defined in a loop are visible outside of it too

Reply:
@apaszke I don't think maintaining compatibility with Python is a good idea. Right now our control flow operators are desugarable into functions which take lambdas as arguments; that is to say, environment semantics apply universally for all nodes which take graph arguments. It makes little sense to introduce special forms to the core language just to emulate Python's broken semantics.

Better is, if you really want the Python style behavior, to desugar a Python style AST into the structured one. That's not too hard, assuming that you can do a little bit of analysis.

Reply:
I'm just pointing that out. I can patch the Python frontend to append such Variables as trailing outputs of loops, but I remember @zdevito saying that the Python thing should be a very dumb translator, that doesn't do any analysis, which means that we might want to stay close to Python semantics.